### PR TITLE
Email Support

### PR DESCRIPTION
**Pull request by liushuyu**
This is a relatively large PR to add email support to LMMS' LSP.

Known issues:

- Currently none

TODOs:

- [x] Better email templates
- [x] Better handling of sending process
- [x] Send the email on registration

Known caveats:

- Potentially destructive to the existing database (requires DB conversion)
- ~~Email server issues unresolved (?)~~ Done.

`More descriptions to come`

---
This change is [Reviewable](https://reviewable.io/reviews/lmms/lmms.io/280)

### CODE REVIEW

File: public/lsp/dbo.php [hunk 1]
```diff
@@ -22,24 +25,24 @@
...
- * Global paths
- */
+* Global paths
+*/
 $TMP_DIR = $_SERVER['DOCUMENT_ROOT'] . '/../tmp/';
 $DATA_DIR = $_SERVER['DOCUMENT_ROOT'] . '/../tmp/';
 $LSP_URL = '/lsp/';
 
 /*
- * By default, the LSP will use the default database values defined above
```
Comment (tresf) [inline public/lsp/dbo.php]: The comment/styling changes seem unwarranted?

File: dev/sql/00-base-schema.sql [hunk 2]
```diff
@@ -0,0 +1,76 @@
+-- Users
```
Comment (tresf) [inline dev/sql/00-base-schema.sql]: Perhaps something indicating that it's for LSP in the filename?

**Comment by tresf** [comment]
@liushuyu I've noticed the composer files are rewritten, I assume that's intentional?  Besides a few cosmetic issues, I don't see any problems.  Thanks for taking the time to document the schema finally.  We need to add this to the wiki too:
* https://github.com/LMMS/lmms.io/wiki/LSP-Installation
* https://github.com/LMMS/lmms.io/wiki/LSP-Database-Management

**Comment by liushuyu** [comment]
>@liushuyu I've noticed the composer files are rewritten, I assume that's intentional? 

Yes, it's intentional, new dependencies are added and `composer` found out some of the existing dependencies needed to upgrade to accommodate the dependency changes. Also, I used some PHP syntax that is from higher versions, so the PHP version required has also been updated.

File: dev/sql/00-base-schema.sql (hunk 2 above)
Comment (liushuyu) [reply dev/sql/00-base-schema.sql]: I will add a description to the `README.md` in the directory (currently it's an empty file)

File: public/lsp/dbo.php (hunk 1 above)
Comment (liushuyu) [reply public/lsp/dbo.php]: Fixed

**Comment by tresf** [comment]
* ✅ I created a user `tresf` in the database for myself that does NOT have an email address.
   ```sql
   insert into users (login, password, pw, realname, is_admin, loginFailureCount)
      values('tresf', 'da1c10a4981fbf545662d36f37a2db9b14f91a6d', '', 'Tres Finocchiaro', 0, 0);
   ```
* ✅ I tried to upload a file, got the error that I need to verify my email first.
* 🚫 I clicked the link to verify email and get this error:
   Screen Shot 2019-09-30 at 2 49 45 PM

File: public/lsp/send_email.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            array("<a href=\"$settings_url\">User Settings</a>"),
+            $settings_url);
+            return;
+        }
+        if (send_email(SESSION()) == true) {
+            display_success("An email with activation link has been sent to your email address.",
+            array("<a href=\"$settings_url\">User Settings</a>"),
+            $settings_url);
+        } else {
+            display_error("Server internal error. Please contact <a href=\"mailto:webmaster@lmms.io" .
```
Comment (tresf) [inline public/lsp/send_email.php]: Why redirect on error?  They might miss the message.

**Comment by liushuyu** [comment]
>🚫 I clicked the link to verify email and get this error:

Fixed

**Comment by tresf** [comment]
Ok, the error is gone, it's just as strange.

1. If the email's empty, why not redirect to the user settings? Instead I have to.. 
   a. Click the link to confirm email, get yelled at for not having an email...
   b. Click the link to go into settings to get an email.<br> It seems like this could just go to settings.  I just don't like the counter at all for errors.  If you can just provide the link and NOT redirect, it puts the power in the user's hands.
2. If I let the counter finish counting now, it just goes back to the upload screen.  If we keep the counter, I feel like it should go to the same place as if the link was clicked.

**Comment by liushuyu** [comment]
> Ok, the error is gone, it's just as strange.
> 
> 1. If the email's empty, why not redirect to the user settings? Instead I have to..
>    
>    * Click the link to confirm email, get yelled at for not having an email...
>    * Click the link to go into settings to get an email. It seems like this could just go to settings.  I just don't like the counter at all for errors.  If you can just provide the link and NOT redirect, it puts the power in the user's hands.
> 2. If I let the counter finish counting now, it just goes back to the upload screen.  If we keep the counter, I feel like it should go to the same place as if the link was clicked.

Okay, now the redirect counter is gone. The link now points to the user settings.

**Comment by tresf** [comment]
Were the changed pushed?  The counter is still there for me.

**Comment by liushuyu** [comment]
>Were the changed pushed? The counter is still there for me.

Yes, the deployment script was stuck. I have manually deployed the site.

**Comment by tresf** [comment]
@liushuyu it works great.  I only have one problem, the email is barely readable
Screen Shot 2019-10-03 at 10 56 37 PM

**Comment by liushuyu** [comment]
>@liushuyu it works great. I only have one problem, the email is barely readable

@Umcaruje said he will do a design pass on the email template shortly

**Comment by tresf** [comment]
@liushuyu ok, hopefully he fixes it soon.  If not, I would recommend a plain text email and he can follow-up with email design later.

File: public/lsp/sidebar.php [hunk 4]
```diff
@@ -76,7 +76,8 @@
 					<input type="hidden" name="subcategory" value="<?php echo GET('subcategory');?>" />
 					<input type="hidden" name="oldaction" value="<?php echo GET('action');?>" />
 					</form>
-					<a href="?action=register"><span class="fas fa-chevron-circle-right"></span>&nbsp;Not registered yet?</a>
+					<a href="?action=register"><span class="fas fa-chevron-circle-right"></span>&nbsp;Not registered yet?</a><br />
+					<a href="?account=forget"><span class="fas fa-chevron-circle-right"></span>&nbsp;Forget password?</a>
```
Comment (Umcaruje) [inline public/lsp/sidebar.php]: I think this should be `Forgot password?`

As per
image
image

**Comment by liushuyu** [comment]
Current known issues are fixed. Testing instance updated, and email looks good even in `neomutt`.

**Comment by liushuyu** [comment]
I am going to halt this experiment as adding features to an already shaky foundation is a really bad idea.

Maybe next time when we have a better infrastructure.

**Comment by tresf** [comment]
Why?  The code was done and the tests passed.  Gathering emails now helps us transition to a new system when we need to.

**Comment by liushuyu** [comment]
>Why? The code was done and the tests passed. Gathering emails now helps us transition to a new system when we need to.

I thought so, the decision was not made lightly. I myself poured much effort into this, the issue with it currently is that there are many vulnerabilities exposed.

I plan to fix those first and then come back to this feature so that user data will be more secure in the storage.

**Pull request closed without merging**